src/t_hash.c

```diff
@@ -185,6 +186,231 @@ static void hashDictWithExpireOnRelease(dict *d) {
     ebDestroy(&dictExpireMeta->hfe, &hashFieldExpireBucketsType, NULL);
 }
 
+#define HASH_LP_NO_TTL 0
+
+typedef struct listpackHfe {
+    ExpireMeta meta;
+    void *lp;
+    sds key;
+} listpackHfe;
+
+int hashLpIsHfe(void *ptr) { return (int) (((uintptr_t) ptr) & 1U); }
+void hashLpSetHfe(void **ptr) { *ptr = (void*) (((uintptr_t) *ptr) | 1); }
+listpackHfe *hashLpGet(void *ptr) {return (void*) ((uintptr_t) ptr & ~1);}
+void *hashLpGetListpack(void *ptr){return hashLpIsHfe(ptr) ? hashLpGet(ptr)->lp : ptr; }
+void hashLpSetListpack(robj *o, unsigned char *lp) {
+    if (hashLpIsHfe(o->ptr))
+        hashLpGet(o->ptr)->lp = lp;
+    else
+        o->ptr = lp;
+}
+
+static uint64_t hashLpExpireDryRun(const robj *o, uint64_t now) {
+    unsigned char *fptr;
+    unsigned char *lp = hashLpGetListpack(o->ptr);
+    uint64_t expired = 0;
+
+    fptr = lpFirst(lp);
+    while (fptr != NULL) {
+        long long val;
+
+        fptr = lpNext(lp, fptr);
+        fptr = lpNext(lp, fptr);
```

Comment:
it  could be corrupt.
```suggestion
        assert(fptr = lpNext(lp, fptr));
        assert(fptr = lpNext(lp, fptr));
```

Reply:
there are some places same as this.